### ISSUE TITLE
Android Firebase Realtime Database  System.currentTimeMillis() value is different on firebase database than generated on app logcat.

### ISSUE THREAD

**Issue by varun7952** [firebase/firebase-android-sdk#3248]
* Android Studio version: Arctic Fox (2020.3.1) 
  * Firebase Component: Realtime Database
  * Component version:  com.google.firebase:firebase-bom:28.0.1'

The problem

In my android app i am submitting the last seen to Realtime Database  ".info/connected" and calling FirebaseDatabase.goOffline(); at onStop and back online by FirebaseDatabase.goOnline(); at onResume. This process work fine every time but when user going offline by calling the listener will call and make user offline with the system timestamp. But when comparing system generate timestamp and logged at server, it always the timestamp which previous timestamp not the same at per the logcat (current timestamp) on android device. 

It's not about the timestamp only, i tried to generate random long, int value but it always the previous generate one.

Use Case: User enter into app and app will fire listener and where we are generating timestamp (for example take random int value = 1) but it won't record on server because user is online. Now when user actually go offline the listener will call and this time app will record the random timestamp/int/long (value = 2 entered) value and save it on server. The value entered at database is the last value not the current one. 

I also tried with new android studio blank project. It create the same issue even on new project.

Code to reproduce the issue.

```
package com.example.myapplication;

import android.os.Bundle;

import com.google.android.material.snackbar.Snackbar;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.text.format.DateFormat;
import android.util.Log;
import android.view.View;

import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;

import com.example.myapplication.databinding.ActivityMainBinding;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import android.view.Menu;
import android.view.MenuItem;

import java.util.Calendar;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "Main Activity Test";
    private AppBarConfiguration appBarConfiguration;
    private ActivityMainBinding binding;
    FirebaseDatabase DatabaseInstance;
    DatabaseReference infoConnected;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setSupportActionBar(binding.toolbar);

        NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment_content_main);
        appBarConfiguration = new AppBarConfiguration.Builder(navController.getGraph()).build();
        NavigationUI.setupActionBarWithNavController(this, navController, appBarConfiguration);
        DatabaseInstance = FirebaseDatabase.getInstance(); // Single Instance
        infoConnected = DatabaseInstance.getReference(".info/connected"); // To check firebase connection state
        initialiseOnlineStatus();
        binding.fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
            }
        });
    }

    @Override
    protected void onPause() {
        super.onPause();
        Log.d(TAG, "onPause: Calling");
    }

    @Override
    protected void onStop() {
        super.onStop();
        Log.d(TAG, "onStop: Calling");
        DatabaseInstance.goOffline();
    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.d(TAG, "onResume: Calling");
        DatabaseInstance.goOnline();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onSupportNavigateUp() {
        NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment_content_main);
        return NavigationUI.navigateUp(navController, appBarConfiguration)
                || super.onSupportNavigateUp();
    }

    // Update Self User Status On Firebase Database
    private void initialiseOnlineStatus() {
        final DatabaseReference Online = DatabaseInstance.getReference("Users/" + "4" + "/Online");
        final DatabaseReference lastOnline = DatabaseInstance.getReference("Users/" + "4" + "/lastSeen");

        infoConnected.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                boolean connected = dataSnapshot.getValue(Boolean.class);
                Log.d(TAG, "onDataChange: Check Connection  Firebase = "+connected);
                if (connected) {
                    Online.setValue(Boolean.TRUE);
                    lastOnline.removeValue();
                    Log.d(TAG, "Connected To Firebase Server ");
                } else {
                    Log.d(TAG, "Disconnect with firebase server ");
                    String offlineTime = String.valueOf(System.currentTimeMillis());
                    Online.onDisconnect().setValue(Boolean.FALSE);
                    Log.d(TAG, "onDataChange: Connected Is False "+connected);
                    Log.d(TAG, "Offline Time " + getLocalDeviceTimestamp(Long.parseLong(offlineTime)) + " " + offlineTime);
                    lastOnline.onDisconnect().setValue(offlineTime);

                }

            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });


    }

    public String getLocalDeviceTimestamp(long msgTimeMillis) {

        Calendar messageTime = Calendar.getInstance();
        messageTime.setTimeInMillis(msgTimeMillis);
        // get Currunt time
        Calendar now = Calendar.getInstance();

        final String strTimeFormate = "h:mm aa";
        final String strDateFormate = "dd/MM/yyyy h:mm aa";
        final String earlierDateFormate = "EEEE dd MMM yyyy, h:mm aa";

        if (now.get(Calendar.DATE) == messageTime.get(Calendar.DATE)
                &&
                ((now.get(Calendar.MONTH) == messageTime.get(Calendar.MONTH)))
                &&
                ((now.get(Calendar.YEAR) == messageTime.get(Calendar.YEAR)))
        ) {

            return "" + DateFormat.format(strTimeFormate, messageTime); // today

        } else if (
                ((now.get(Calendar.DATE) - messageTime.get(Calendar.DATE)) == 1)
                        &&
                        ((now.get(Calendar.MONTH) == messageTime.get(Calendar.MONTH)))
                        &&
                        ((now.get(Calendar.YEAR) == messageTime.get(Calendar.YEAR)))
        ) {
            return "yesterday " + DateFormat.format(strTimeFormate, messageTime); // yesterday
        } else {
            return "" + DateFormat.format(earlierDateFormate, messageTime); // earlier
        }
    }


}
```

**Comment by puf** [comment]
As I said on the Stack Overflow [post](https://stackoverflow.com/questions/70323201/android-system-currenttimemillis-value-is-different-on-firebase-database-than/70325092?noredirect=1#comment124447706_70325092) where we've been discussing this, I doubt this is a bug in the SDK or server as neither of them can be generating these "wrong" values that you see.

Most likely you're registering multiple `onDisconnect` handlers, which makes it hard to predict which one will fire first/last. The solution is to ensure you register only one `onDisconnect` handler for each path in your database.

**Comment by varun7952** [comment]
> As I said on the Stack Overflow [post](https://stackoverflow.com/questions/70323201/android-system-currenttimemillis-value-is-different-on-firebase-database-than/70325092?noredirect=1#comment124447706_70325092) where we've been discussing this, I doubt this is a bug in the SDK or server as neither of them can be generating these "wrong" values that you see.
> 
> Most likely you're registering multiple `onDisconnect` handlers, which makes it hard to predict which one will fire first/last. The solution is to ensure you register only one `onDisconnect` handler for each path in your database.

@puf You can see by yourself there is no multiple onDisconnect in code. You can reproduce the error and I'm repeatedly say it's a complete code where you can reproduce the issue. 

You can copy and paste same code to get the issue on your end

**Comment by puf** [comment]
Can you show the logcat output of your code? Specifically, I suspect that this shows multiple times `Log.d(TAG, "onDataChange: Check Connection  Firebase = "+connected)` due to the way the connection is established.

**Comment by varun7952** [comment]
> Can you show the logcat output of your code? Specifically, I suspect that this shows multiple times `Log.d(TAG, "onDataChange: Check Connection Firebase = "+connected)` due to the way the connection is established.

Sure I will share with you soon and that's also i mentioned at SO that it calls multiple times.

**Comment by varun7952** [comment]
> Can you show the logcat output of your code? Specifically, I suspect that this shows multiple times `Log.d(TAG, "onDataChange: Check Connection Firebase = "+connected)` due to the way the connection is established.

Please Check the complete log from app closed (cleared from recent too ) then run then close again by pressing Home button.

```
2021-12-19 01:40:17.658 20837-20837/? D/ActivityThread: setConscryptValidator
2021-12-19 01:40:17.659 20837-20837/? D/ActivityThread: setConscryptValidator - put
2021-12-19 01:40:17.675 20837-20837/? D/ActivityThread: handleBindApplication()++ app=com.example.myapplication
2021-12-19 01:40:17.698 20837-20837/com.example.myapplication D/ActivityThread: handleBindApplication() -- skipGraphicsSupport=false
2021-12-19 01:40:17.701 20837-20837/com.example.myapplication D/ActivityThread: ActivityThread::handleMakeApplication() data=AppBindData{appInfo=ApplicationInfo{7ad8b90 com.example.myapplication}} appContext=android.app.ContextImpl@6ca7c89 appContext.mOpPackageName=com.example.myapplication appContext.mBasePackageName=com.example.myapplication appContext.mPackageInfo=android.app.LoadedApk@11ef48e data.restrictedBackupMode= false data.providers= [ContentProviderInfo{name=com.example.myapplication.firebaseinitprovider className=com.google.firebase.provider.FirebaseInitProvider}] Caller=android.app.ActivityThread.handleBindApplication:7440 android.app.ActivityThread.access$1400:301 android.app.ActivityThread$H.handleMessage:2148 android.os.Handler.dispatchMessage:106 android.os.Looper.loop:246 
2021-12-19 01:40:17.702 20837-20837/com.example.myapplication D/LoadedApk: LoadedApk::makeApplication() appContext=android.app.ContextImpl@c8594af appContext.mOpPackageName=com.example.myapplication appContext.mBasePackageName=com.example.myapplication appContext.mPackageInfo=android.app.LoadedApk@11ef48e
2021-12-19 01:40:17.702 20837-20837/com.example.myapplication D/NetworkSecurityConfig: No Network Security Config specified, using platform default
2021-12-19 01:40:17.724 20837-20837/com.example.myapplication D/NetworkSecurityConfig: No Network Security Config specified, using platform default
2021-12-19 01:40:17.734 20837-20837/com.example.myapplication I/FirebaseApp: Device unlocked: initializing all Firebase APIs for app [DEFAULT]
2021-12-19 01:40:17.734 20837-20837/com.example.myapplication I/FirebaseInitProvider: FirebaseApp initialization successful
2021-12-19 01:40:17.797 20837-21250/com.example.myapplication I/AdrenoGLES-0: QUALCOMM build                   : e1ac91e, I2b3b5fbd00
    Build Date                       : 11/17/20
    OpenGL ES Shader Compiler Version: EV031.32.02.06
    Local Branch                     : 
    Remote Branch                    : 
    Remote Branch                    : 
    Reconstruct Branch               : 
2021-12-19 01:40:17.797 20837-21250/com.example.myapplication I/AdrenoGLES-0: Build Config                     : S P 10.0.7 AArch64
2021-12-19 01:40:17.797 20837-21250/com.example.myapplication I/AdrenoGLES-0: Driver Path                      : /vendor/lib64/egl/libGLESv2_adreno.so
2021-12-19 01:40:17.806 20837-21250/com.example.myapplication I/AdrenoGLES-0: PFP: 0x016ee190, ME: 0x00000000
2021-12-19 01:40:17.949 20837-20837/com.example.myapplication I/DecorView: [INFO] isPopOver=false, config=true
2021-12-19 01:40:17.950 20837-20837/com.example.myapplication I/DecorView: updateCaptionType >> DecorView@75a69a[], isFloating=false, isApplication=true, hasWindowDecorCaption=false, hasWindowControllerCallback=true
2021-12-19 01:40:17.950 20837-20837/com.example.myapplication D/DecorView: setCaptionType = 0, this = DecorView@75a69a[]
2021-12-19 01:40:17.961 20837-20837/com.example.myapplication W/e.myapplicatio: Accessing hidden method Landroid/view/View;->computeFitSystemWindows(Landroid/graphics/Rect;Landroid/graphics/Rect;)Z (greylist, reflection, allowed)
2021-12-19 01:40:17.961 20837-20837/com.example.myapplication W/e.myapplicatio: Accessing hidden method Landroid/view/ViewGroup;->makeOptionalFitsSystemWindows()V (greylist, reflection, allowed)
2021-12-19 01:40:18.001 20837-20837/com.example.myapplication D/InputTransport: Input channel constructed: '56debfe', fd=79
2021-12-19 01:40:18.003 20837-20837/com.example.myapplication I/ViewRootImpl@ed0b127[MainActivity]: setView = com.android.internal.policy.DecorView@75a69a TM=true
2021-12-19 01:40:18.006 20837-20837/com.example.myapplication D/Main Activity Test: onDataChange: Check Connection  Firebase = false
2021-12-19 01:40:18.006 20837-20837/com.example.myapplication D/Main Activity Test: Disconnect with firebase server 
2021-12-19 01:40:18.006 20837-20837/com.example.myapplication D/Main Activity Test: onDataChange: Connected Is False false
2021-12-19 01:40:18.007 20837-20837/com.example.myapplication D/Main Activity Test: Offline Time 1:40 am 1639858218006
2021-12-19 01:40:18.050 20837-20837/com.example.myapplication I/SurfaceControl: assignNativeObject: nativeObject = 0 Surface(name=null)/@0x428d26e / android.view.SurfaceControl.readFromParcel:1115 android.view.IWindowSession$Stub$Proxy.relayout:1787 android.view.ViewRootImpl.relayoutWindow:8953 android.view.ViewRootImpl.performTraversals:3312 android.view.ViewRootImpl.doTraversal:2575 android.view.ViewRootImpl$TraversalRunnable.run:9910 android.view.Choreographer$CallbackRecord.run:1010 android.view.Choreographer.doCallbacks:809 android.view.Choreographer.doFrame:744 android.view.Choreographer$FrameDisplayEventReceiver.run:995 
2021-12-19 01:40:18.052 20837-20837/com.example.myapplication I/ViewRootImpl@ed0b127[MainActivity]: Relayout returned: old=(0,0,1080,2400) new=(0,0,1080,2400) req=(1080,2400)0 dur=13 res=0x7 s={true -5476376669637074272} ch=true fn=-1
2021-12-19 01:40:18.130 20837-20837/com.example.myapplication I/ViewRootImpl@ed0b127[MainActivity]: MSG_WINDOW_FOCUS_CHANGED 1 1
2021-12-19 01:40:18.131 20837-20837/com.example.myapplication D/InputMethodManager: prepareNavigationBarInfo() DecorView@75a69a[MainActivity]
2021-12-19 01:40:18.131 20837-20837/com.example.myapplication D/InputMethodManager: getNavigationBarColor() -16711423
2021-12-19 01:40:18.133 20837-20837/com.example.myapplication D/InputMethodManager: prepareNavigationBarInfo() DecorView@75a69a[MainActivity]
2021-12-19 01:40:18.133 20837-20837/com.example.myapplication D/InputMethodManager: getNavigationBarColor() -16711423
2021-12-19 01:40:18.133 20837-20837/com.example.myapplication D/InputMethodManager: startInputInner - Id : 0
2021-12-19 01:40:18.133 20837-20837/com.example.myapplication I/InputMethodManager: startInputInner - mService.startInputOrWindowGainedFocus
2021-12-19 01:40:18.138 20837-21243/com.example.myapplication D/InputTransport: Input channel constructed: 'ClientS', fd=87
2021-12-19 01:40:18.139 20837-20837/com.example.myapplication D/InputMethodManager: prepareNavigationBarInfo() DecorView@75a69a[MainActivity]
2021-12-19 01:40:18.139 20837-20837/com.example.myapplication D/InputMethodManager: getNavigationBarColor() -16711423
2021-12-19 01:40:18.139 20837-20837/com.example.myapplication D/InputMethodManager: startInputInner - Id : 0
2021-12-19 01:40:18.148 20837-20837/com.example.myapplication D/SurfaceControl: hide : mNativeObject = -5476376673931944736 - sc.mNativeObject = -5476376673931640608 - Surface(name=Surface(name=a90d009 InputMethod)/@0xc92da3c - animation-leash)/@0x7c1ab2b
2021-12-19 01:40:18.148 20837-20837/com.example.myapplication D/SurfaceControl: nativeSetFlags Done : Surface(name=Surface(name=a90d009 InputMethod)/@0xc92da3c - animation-leash)/@0x7c1ab2b
2021-12-19 01:40:18.149 20837-20837/com.example.myapplication I/SurfaceControl: release : mNativeObject = -5476376673931985920 - Surface(name=Surface(name=c81cbbb NavigationBar0)/@0x82ee257 - animation-leash)/@0xddf2b88 / android.view.-$$Lambda$Rl1VZmNJ0VZDLK0BAbaVGis0rrA.accept:2 android.view.InsetsSourceControl.release:170 android.view.InsetsSourceConsumer.setControl:202 android.view.InsetsController.onControlsChanged:833 
2021-12-19 01:40:18.149 20837-20837/com.example.myapplication I/SurfaceControl: nativeRelease nativeObject s[-5476376673931985920]
2021-12-19 01:40:18.149 20837-20837/com.example.myapplication I/SurfaceControl: nativeRelease nativeObject e[-5476376673931985920]
2021-12-19 01:40:18.150 20837-20837/com.example.myapplication I/SurfaceControl: release : mNativeObject = -5476376673931984160 - Surface(name=Surface(name=f7e0f52 StatusBar)/@0x3765975 - animation-leash)/@0xa792921 / android.view.-$$Lambda$Rl1VZmNJ0VZDLK0BAbaVGis0rrA.accept:2 android.view.InsetsSourceControl.release:170 android.view.InsetsSourceConsumer.setControl:202 android.view.InsetsController.onControlsChanged:833 
2021-12-19 01:40:18.150 20837-20837/com.example.myapplication I/SurfaceControl: nativeRelease nativeObject s[-5476376673931984160]
2021-12-19 01:40:18.150 20837-20837/com.example.myapplication I/SurfaceControl: nativeRelease nativeObject e[-5476376673931984160]
2021-12-19 01:40:19.073 20837-20837/com.example.myapplication D/Main Activity Test: onDataChange: Check Connection  Firebase = true
2021-12-19 01:40:19.074 20837-20837/com.example.myapplication D/Main Activity Test: Connected To Firebase Server 
2021-12-19 01:40:22.811 20837-20837/com.example.myapplication I/ViewRootImpl@ed0b127[MainActivity]: MSG_WINDOW_FOCUS_CHANGED 0 1
2021-12-19 01:40:22.862 20837-20837/com.example.myapplication I/SurfaceControl: release : mNativeObject = -5476376673931640608 - Surface(name=Surface(name=a90d009 InputMethod)/@0xc92da3c - animation-leash)/@0x7c1ab2b / android.view.-$$Lambda$Rl1VZmNJ0VZDLK0BAbaVGis0rrA.accept:2 android.view.InsetsSourceControl.release:170 android.view.InsetsSourceConsumer.setControl:202 android.view.ImeInsetsSourceConsumer.setControl:154 
2021-12-19 01:40:22.862 20837-20837/com.example.myapplication I/SurfaceControl: nativeRelease nativeObject s[-5476376673931640608]
2021-12-19 01:40:22.862 20837-20837/com.example.myapplication I/SurfaceControl: nativeRelease nativeObject e[-5476376673931640608]
2021-12-19 01:40:22.862 20837-20837/com.example.myapplication I/SurfaceControl: release : mNativeObject = -5476376673931638672 - Surface(name=Surface(name=c81cbbb NavigationBar0)/@0x82ee257 - animation-leash)/@0xa90dfc / android.view.-$$Lambda$Rl1VZmNJ0VZDLK0BAbaVGis0rrA.accept:2 android.view.InsetsSourceControl.release:170 android.view.InsetsSourceConsumer.setControl:202 android.view.InsetsController.onControlsChanged:833 
2021-12-19 01:40:22.870 20837-20837/com.example.myapplication I/SurfaceControl: nativeRelease nativeObject s[-5476376673931638672]
2021-12-19 01:40:22.870 20837-20837/com.example.myapplication I/SurfaceControl: nativeRelease nativeObject e[-5476376673931638672]
2021-12-19 01:40:22.870 20837-20837/com.example.myapplication I/SurfaceControl: release : mNativeObject = -5476376673931635328 - Surface(name=Surface(name=f7e0f52 StatusBar)/@0x3765975 - animation-leash)/@0x725ec85 / android.view.-$$Lambda$Rl1VZmNJ0VZDLK0BAbaVGis0rrA.accept:2 android.view.InsetsSourceControl.release:170 android.view.InsetsSourceConsumer.setControl:202 android.view.InsetsController.onControlsChanged:833 
2021-12-19 01:40:22.870 20837-20837/com.example.myapplication I/SurfaceControl: nativeRelease nativeObject s[-5476376673931635328]
2021-12-19 01:40:22.870 20837-20837/com.example.myapplication I/SurfaceControl: nativeRelease nativeObject e[-5476376673931635328]
2021-12-19 01:40:22.945 20837-20837/com.example.myapplication D/InputTransport: Input channel destroyed: 'ClientS', fd=87
2021-12-19 01:40:22.957 20837-20837/com.example.myapplication I/SurfaceControl: release : mNativeObject = -5476376673931628640 - Surface(name=Surface(name=a90d009 InputMethod)/@0xc92da3c - animation-leash)/@0xca32094 / android.view.-$$Lambda$Rl1VZmNJ0VZDLK0BAbaVGis0rrA.accept:2 android.view.InsetsSourceControl.release:170 android.view.InsetsSourceConsumer.setControl:202 android.view.ImeInsetsSourceConsumer.setControl:154 
2021-12-19 01:40:22.958 20837-20837/com.example.myapplication I/SurfaceControl: nativeRelease nativeObject s[-5476376673931628640]
2021-12-19 01:40:22.958 20837-20837/com.example.myapplication I/SurfaceControl: nativeRelease nativeObject e[-5476376673931628640]
2021-12-19 01:40:23.188 20837-20837/com.example.myapplication I/SurfaceControl: release : mNativeObject = -5476376673931961104 - Surface(name=null)/@0xbda273d / android.view.SurfaceControl.assignNativeObject:567 android.view.SurfaceControl.readFromParcel:1115 android.view.IWindowSession$Stub$Proxy.relayout:1777 android.view.ViewRootImpl.relayoutWindow:8953 
2021-12-19 01:40:23.188 20837-20837/com.example.myapplication I/SurfaceControl: nativeRelease nativeObject s[-5476376673931961104]
2021-12-19 01:40:23.188 20837-20837/com.example.myapplication I/SurfaceControl: nativeRelease nativeObject e[-5476376673931961104]
2021-12-19 01:40:23.189 20837-20837/com.example.myapplication I/SurfaceControl: assignNativeObject: nativeObject = 0 Surface(name=null)/@0xbda273d / android.view.SurfaceControl.readFromParcel:1115 android.view.IWindowSession$Stub$Proxy.relayout:1777 android.view.ViewRootImpl.relayoutWindow:8953 android.view.ViewRootImpl.performTraversals:3312 android.view.ViewRootImpl.doTraversal:2575 android.view.ViewRootImpl$TraversalRunnable.run:9910 android.view.Choreographer$CallbackRecord.run:1010 android.view.Choreographer.doCallbacks:809 android.view.Choreographer.doFrame:744 android.view.Choreographer$FrameDisplayEventReceiver.run:995 
2021-12-19 01:40:23.190 20837-20837/com.example.myapplication I/SurfaceControl: assignNativeObject: nativeObject = 0 Surface(name=null)/@0x428d26e / android.view.SurfaceControl.readFromParcel:1115 android.view.IWindowSession$Stub$Proxy.relayout:1787 android.view.ViewRootImpl.relayoutWindow:8953 android.view.ViewRootImpl.performTraversals:3312 android.view.ViewRootImpl.doTraversal:2575 android.view.ViewRootImpl$TraversalRunnable.run:9910 android.view.Choreographer$CallbackRecord.run:1010 android.view.Choreographer.doCallbacks:809 android.view.Choreographer.doFrame:744 android.view.Choreographer$FrameDisplayEventReceiver.run:995 
2021-12-19 01:40:23.192 20837-20837/com.example.myapplication I/ViewRootImpl@ed0b127[MainActivity]: Relayout returned: old=(0,0,1080,2400) new=(0,0,1080,2400) req=(1080,2400)8 dur=16 res=0x5 s={false 0} ch=true fn=6
2021-12-19 01:40:23.196 20837-20837/com.example.myapplication I/ViewRootImpl@ed0b127[MainActivity]: stopped(true) old=false

```

**Comment by varun7952** [comment]
More Clean Log 


```
2021-12-19 04:00:16.516 29264-29264/com.example.myapplication D/Main Activity Test: onResume: Calling
2021-12-19 04:00:16.541 29264-29264/com.example.myapplication D/Main Activity Test: onDataChange: Check Connection  Firebase = false
2021-12-19 04:00:16.541 29264-29264/com.example.myapplication D/Main Activity Test: Disconnect with firebase server 
2021-12-19 04:00:16.542 29264-29264/com.example.myapplication D/Main Activity Test: onDataChange: Connected Is False false
2021-12-19 04:00:16.543 29264-29264/com.example.myapplication D/Main Activity Test: Offline Time 4:00 am 1639866616541
2021-12-19 04:00:16.516 29264-29264/com.example.myapplication D/Main Activity Test: onResume: Calling
2021-12-19 04:00:16.541 29264-29264/com.example.myapplication D/Main Activity Test: onDataChange: Check Connection  Firebase = false
2021-12-19 04:00:16.541 29264-29264/com.example.myapplication D/Main Activity Test: Disconnect with firebase server 
2021-12-19 04:00:16.542 29264-29264/com.example.myapplication D/Main Activity Test: onDataChange: Connected Is False false
2021-12-19 04:00:16.543 29264-29264/com.example.myapplication D/Main Activity Test: Offline Time 4:00 am 1639866616541
2021-12-19 04:00:17.535 29264-29264/com.example.myapplication D/Main Activity Test: onDataChange: Check Connection  Firebase = true
2021-12-19 04:00:17.536 29264-29264/com.example.myapplication D/Main Activity Test: Connected To Firebase Server 
2021-12-19 04:00:52.609 29264-29264/com.example.myapplication D/Main Activity Test: onPause: Calling
2021-12-19 04:00:53.167 29264-29264/com.example.myapplication D/Main Activity Test: onStop: Calling
2021-12-19 04:00:53.205 29264-29264/com.example.myapplication D/Main Activity Test: onDataChange: Check Connection  Firebase = false
2021-12-19 04:00:53.205 29264-29264/com.example.myapplication D/Main Activity Test: Disconnect with firebase server 
2021-12-19 04:00:53.205 29264-29264/com.example.myapplication D/Main Activity Test: onDataChange: Connected Is False false
2021-12-19 04:00:53.208 29264-29264/com.example.myapplication D/Main Activity Test: Offline Time 4:00 am 1639866653205
```

**Comment by puf** [comment]
That seems logical: once the client is disconnected from the server, it can't attach an `onDisconnect` handler on the server anymore.

**Comment by varun7952** [comment]
> That seems logical: once the client is disconnected from the server, it can't attach an `onDisconnect` handler on the server anymore.

You right but when activity onStop called the Database.goOffline will onDataChange then as you can see everything is working fine even online status got updated to false and handler calls for update data. But the value it takes from previous generated timestamp. Is anything wrong with the code even it's not much into code?

**Comment by puf** [comment]
My guess is that the `onDisconnect` handler that you set when `.info/connected` becomes false, will still be sent to the server once you reconnect. And I don't think there's any defined order when you have multiple `onDisconnect` handlers at a single path.

If you look at the sample of a [presence system](https://firebase.google.com/docs/database/web/offline-capabilities#section-sample) in the documentation, you'll note that it generates a new push path every time `.info/connected` becomes true. This is because a single device can have multiple connections over its lifecycle, and those connections may expire at different intervals on the server. Trying to use a single value for this is bound to lead to issues with either marking the user as offline prematurely, or (such as in your case) seeing older values being written.

**Comment by varun7952** [comment]
> 

This is only single activity in my code and no other fragment or activities i have. 

> This is because a single device can have multiple connections over its lifecycle, and those connections may expire at different intervals on the server. Trying to use a single value for this is bound to lead to issues with either marking the user as offline prematurely, or (such as in your case) seeing older values being written.

I don't understand what do you mean by single device can have multiple connections over its lifecycle. If you talking about the code/project i have, for the sake of issue i created this project and i get the same problem, i am using single instance of .info/connected.  I also tried with push but same issue is there. Can you please the code on your side so you can see the issue rather then assuming my code.

**Comment by argzdev** [comment]
Hi @varun7952, thanks for reporting.

If you look at the link(for javascript) @puf has provided. It seems that the correct way of storing timestamp is different from your implementation, I believe this might've been addressed by our engineers in the past. You could try this [link](https://firebase.google.com/docs/database/android/offline-capabilities#server-timestamps) for the android implementation, I've tried to use the same solution and it works. 

TLDR: Instead of using `System.currentTimeMillis()` use `ServerValue.TIMESTAMP` instead. Though there is still a slight difference, but this is relatively small in milliseconds.

Relevant code:
```
userLastOnlineRef.onDisconnect().setValue(ServerValue.TIMESTAMP);
```

You can place that line in either `if(connected)` or in the `else` statement, both ways will work. But I would still recommend moving that line inside `if(connected)` similar to the implementation on the said link.

With that said, I'll be closing this issue now, since the solution for this already exists.

**Comment by varun7952** [comment]
if i use `userLastOnlineRef.onDisconnect().setValue(ServerValue.TIMESTAMP);` then it break the logic of any chat app because that will record the server timestamp no the system timestamp , if one person in US and other is europe then it will be make lot of issues in showing the local timestamp of user.

**Comment by puf** [comment]
If you want to show the local timestamp when each client was online, you will have to periodically write that timestamp while the connection exists. There is no way to use an `onDisconnect()` handler for that, as the value for such a handler has to be known when you register the handler - which you can't do after `.info/connected` has become false.

**Comment by argzdev** [comment]
Hi @varun7952, the reason this doesn't work is because you're trying to set a value after you've already disconnected the connection, it's bound to cause issues.

Your use case:
1. The app goes to background
2. RTDB connection goes offline and disconnects
3. Changing server value after the connection is already disconnected <--- this won't work, afaik if it does then that is a bug.

This is an intended behavior.

What you can do is change the logic of how you store the timestamp. Simply store the timestamp BEFORE you severe the connection. There shouldn't be any difference in time, even if there was, it would only be miniscule.

Relevant code:
```
@Override
    protected void onResume() {
        super.onResume();
        Log.d(TAG, "On Resume Running");

        Online.setValue(Boolean.TRUE);
        lastOnline.removeValue();
        DatabaseInstance.goOnline();
    }

@Override
    protected void onStop() {
        super.onStop();
        Log.d(TAG, "ON STOP CALLING");

        Online.setValue(Boolean.FALSE);
        lastOnline.setValue("" + System.currentTimeMillis());
        DatabaseInstance.goOffline();
    }
```

Then you can simply remove the `onDisconnect` in your ValueEventListener.

**Comment by varun7952** [comment]
@argzdev Thanks alot for this answer. It works by your given example.  But my question is when we call `goOffline` it calls `onDataChange` and also setting boolean value to false and even timestamp value is wrong but it also setting the value for the last seen. My question is when it's work as expected and the `onDataChange` also called then it;s not setting the correct value for the timestamp. Doesn't `goOffline` calls `OnDisconnect` too and setting the values when disconnect?

Btw your code is working as expected.Thanks

**Comment by argzdev** [comment]
I'm glad that works for you. 
As for your question, you've might've missed @puf's reply, it actually explains it very well.

> as the value for such a handler has to be **known** when you register the handler - which you can't do after `.info/connected` has become false.

The keyword here is **known**, this means that the values you insert in `onDisconnect().setValue()` are already read and stored right at the beginning of the app, it isn't stored when your app goes offline. The value cannot be somewhat dynamic that is only read when the app is about to go offline.

You'll notice that in your code, the timestamp that is being stored is the time when your app first went online, this shows that the initial value is being stored then later sent to the server as soon as the app goes offline.

Sample use case:
1. App opens the first time at 10:00:01 pm
2. App automatically calls `onDisconnect.setValue(System.currentTimeMillis())` to store 10:00:01pm
3. App goes to background at 10:01:30 pm
4. App triggers `onDisconnect` to send stored data which is 10:00:01 pm and not 10:01:30 pm.
 
You could also check out the [documentation](https://firebase.google.com/docs/reference/android/com/google/firebase/database/OnDisconnect#public-taskvoid-setvalue-object-value), where it says 

> Ensure the data at this location is set to the "specified" value when the client is disconnected

I hope that clears it up for you, @varun7952.
